episodes/fig/source/06-blurring/create_blur_animation.py

```diff
@@ -0,0 +1,160 @@
+from matplotlib import patches as p
+from matplotlib.animation import FuncAnimation
+from tqdm import tqdm
+
+# Path to input and output images
+data_path = "../../../data/"
+fig_path = "../../../fig/"
+input_file = data_path + "letterA.tif"
+output_file = fig_path + "blur-demo.gif"
+
+# Change here colors to improve accessibility
+kernel_color = "tab:red"
+center_color = "tab:olive"
+kernel_size = 3
+
+### ANIMATION FUNCTIONS
+def init():
+    """
+    Initialization function
+    - Set image array data
+    - Autoscale image display
+    - Set XY coordinates of rectangular patches
+    """
+    im.set_array(img_convolved)
+    im.autoscale()
+    k_rect.set_xy((-0.5, -0.5))
+    c_rect1.set_xy((kernel_size / 2 - 1, kernel_size / 2 - 1))
+    return [im, k_rect, c_rect1]
+
+def update(frame):
+    """
+    Animation update function. For every frame do the following:
+    - Update X and Y coordinates of rectangular patch for kernel
+    - Update X and Y coordinates of rectangular patch for central pixel
+    - Update blurred image frame
+    """
+    pbar.update(1)
+    row = (frame % total_frames) // (img_pad.shape[0] - kernel_size + 1)
+    col = (frame % total_frames) % (img_pad.shape[1] - kernel_size + 1)
+
+    k_rect.set_x(col - 0.5)
+    c_rect1.set_x(col + (kernel_size/2 - 1))
+    k_rect.set_y(row - 0.5)
+    c_rect1.set_y(row + (kernel_size/2 - 1))
+
+    im.set_array(all_frames[frame])
+    im.autoscale()
+
+    return [im, k_rect, c_rect1]
+
+# MAIN PROGRAM
+if __name__ == "__main__":
+
+    print("Creating blurred animation with kernel size:", kernel_size)
+
+    # Load image
+    img = plt.imread(input_file)
+
+    ### HERE WE USE THE CONVOLVE FUNCTION TO GET THE FINAL BLURRED IMAGE
+    # I chose a simple mean filter (equal kernel weights)
+    kernel = np.ones(shape=(kernel_size, kernel_size)) / kernel_size ** 2 # create kernel
+    # convolve the image i.e. apply mean filter
+    img_convolved = convolve(img, kernel, mode='constant', cval=0) # pad borders with zero like below for consistency
+
+
+    ### HERE WE CONVOLVE MANUALLY STEP-BY-STEP TO CREATE ANIMATION
+    img_pad = np.pad(img, (int(np.ceil(kernel_size/2) - 1), int(np.ceil(kernel_size/2) - 1))) # Pad image to deal with borders
+    new_img = np.zeros(img.shape, dtype=np.uint16) # this will be the blurred final image
+
+    # add first frame with complete blurred image for print version of GIF
+    all_frames = [img_convolved]
+
+    # precompute animation frames and append to the list
+    total_frames = (img_pad.shape[0] - kernel_size + 1) * (img_pad.shape[1] - kernel_size + 1) # total frames if by chance image is not squared
+    for frame in range(total_frames):
+        row = (frame % total_frames) // (img_pad.shape[0] - kernel_size + 1) # row index
+        col = (frame % total_frames) % (img_pad.shape[1] - kernel_size + 1) # col index
+        img_chunk = img_pad[row : row + kernel_size, col : col + kernel_size] # get current image chunk inside the kernel
```

Comment:
```suggestion
        img_chunk = img_pad[row:row + kernel_size, col:col + kernel_size]  # get current image chunk inside the kernel
```
I know the spacing is not 'logical' (feel free to add brackets around `row + kernel_size` if you like).